add -image locator strategy and related logic

jlipps opened the pull request:
First stab at putting the find by image logic into base driver.

I know there's no tests and docstrings, will add those. Just wanted to get the ideas here for any initial comments.

I haven't yet done extensive testing, but the happy path works for me locally.

mykola-mokhnach commented on lib/basedriver/commands/find.js:
```diff
@@ -1,8 +1,17 @@
 import log from '../logger';
+import Jimp from 'jimp';
+import Promise from 'bluebird';
```
We usually call it `B`

mykola-mokhnach commented on lib/basedriver/commands/find.js:
```diff
@@ -1,8 +1,17 @@
+import Jimp from 'jimp';
+import Promise from 'bluebird';
+import { errors } from '../../..';
+import { Buffer } from 'buffer';
```
should this be imported explicitly?

mykola-mokhnach commented on lib/basedriver/commands/find.js:
```diff
@@ -32,6 +41,15 @@ helpers.findElOrElsWithProcessing = async function (strategy, selector, mult, co
+    // findElement(s)FromElement), this doesn't make sense, so use of the
+    // -image locator strategy won't pass the strategy validation filter as is
+    // desired
+    return await this.findByImage(selector);
```
I would say there's still a difference, because in case of multiple elements we need to return an empty array or throw an exception if only one element is expected

imurchie replied:
But in case of multiple elements (or anything other than strict `findElement`) it will throw an error saying there is no such strategy (https://github.com/appium/appium-base-driver/blob/master/lib/basedriver/commands/find.js#L19).

mykola-mokhnach commented on lib/basedriver/commands/find.js:
```diff
@@ -47,6 +65,120 @@ commands.findElementsFromElement = async function (strategy, selector, elementId
   return await this.findElOrElsWithProcessing(strategy, selector, true, elementId);
 };
 
+helpers.findByImage = async function (b64Template, {checkingStaleness = false} = {}) {
```
-> shouldCheckStaleness

mykola-mokhnach commented on lib/basedriver/commands/find.js:
```diff
@@ -47,6 +65,120 @@ commands.findElementsFromElement = async function (strategy, selector, elementId
+helpers.findByImage = async function (b64Template, {checkingStaleness = false} = {}) {
+  let threshold = DEFAULT_MATCH_THRESHOLD;
+  if (this.settings) {
+    threshold = this.settings.getSettings().imageMatchThreshold;
```
shouldn't we only assign it if `imageMatchThreshold` is defined?

mykola-mokhnach commented on lib/basedriver/commands/find.js:
```diff
@@ -47,6 +65,120 @@ commands.findElementsFromElement = async function (strategy, selector, elementId
+  // someone might have sent in a template that's larger than the screen
+  // dimensions. If so let's check and cut it down to size since the algorithm
+  // will not work unless we do
+  b64Template = await this.ensureTemplateSize(b64Template, screenWidth, screenHeight);
```
I would rather skip this check and let the client to be responsible for template size. It might be too expensive to call `getWindowSize` each time the lookup is executed.

mykola-mokhnach commented on lib/basedriver/commands/find.js:
```diff
@@ -47,6 +65,120 @@ commands.findElementsFromElement = async function (strategy, selector, elementId
+      return true;
+    });
+  } catch (err) {
+    if (!err.match(/Condition unmet/)) {
```
It might be handy to have some explanation here

mykola-mokhnach commented on lib/basedriver/commands/find.js:
```diff
@@ -47,6 +65,120 @@ commands.findElementsFromElement = async function (strategy, selector, elementId
+  // are two potential types of mismatch: aspect ratio mismatch and scale
+  // mismatch. we need to detect and fix both
+
+  const screenAR = screenWidth / screenHeight;
```
screen aspect ration is not going to be changed during the session, so it might be safe to cache it

jlipps replied:
i think it could be changed during session if the device orientation is changed

mykola-mokhnach commented on lib/basedriver/driver.js:
```diff
@@ -276,6 +280,11 @@ class BaseDriver extends Protocol {
         // outside, rejecting the current command immediately
         this.curCommandCancellable = new B((r) => { r(); }).then(() => { // eslint-disable-line promise/prefer-await-to-then
+          // if one of the args is an image element, handle it separately
+          const imgElId = getImgElFromArgs(args);
```
this looks a bit hacky

jlipps replied:
hmm, it was the only way I could think of to pass control to ImageElement in the case of an image element existing in the arguments. I agree it's not super pretty. Any other ideas?

mykola-mokhnach commented on lib/basedriver/image-element.js:
```diff
@@ -0,0 +1,169 @@
+      throw new errors.StaleElementReferenceError();
+    }
+
+    if (!this.equals(newImgEl)) {
```
I'm not quite sure if `click` should mutate the element

mykola-mokhnach commented on lib/basedriver/commands/find.js:
```diff
@@ -47,6 +65,120 @@ commands.findElementsFromElement = async function (strategy, selector, elementId
+    return b64Screenshot;
+  }
+
+  // otherwise, if they don't match, it could spell problems for the accuracy
```
I would rather make this configurable from settings to make it as performant as possible by default and only do these transformations if explicitly asked by the client

jlipps commented:
(note that this PR now depends on https://github.com/appium/appium-support/pull/77)

mykola-mokhnach commented on lib/basedriver/commands/find.js:
```diff
@@ -111,7 +110,8 @@ helpers.findByImage = async function (b64Template, {checkingStaleness = false} =
 
 helpers.ensureTemplateSize = async function (b64Template, screenWidth, screenHeight) {
-  let {width: tplWidth, height: tplHeight, imgObj} = await getJimpData(b64Template);
+  let imgObj = await imageUtils.getJimpData(b64Template);
```
is the function called `getJimpImage` or `getJimpData` ?

jlipps replied:
oops, fixed.

mykola-mokhnach commented on lib/basedriver/commands/find.js:
```diff
@@ -64,18 +63,40 @@ commands.findElementsFromElement = async function (strategy, selector, elementId
-  }
+/**
+ * @typedef {Object} FindByImageOptions
+ * @property {boolean} shouldCheckStaleness - whether this call to find an
```
can we also mention the default values for these properties or if any of them is required to set?

mykola-mokhnach commented on lib/basedriver/commands/find.js:
```diff
@@ -64,18 +63,40 @@ commands.findElementsFromElement = async function (strategy, selector, elementId
+helpers.findByImage = async function (b64Template, {
+  shouldCheckStaleness = false,
+  multiple = false,
+} = {}) {
```
I assume `= {}` is redundant

mykola-mokhnach commented on lib/basedriver/commands/find.js:
```diff
@@ -64,18 +63,40 @@ commands.findElementsFromElement = async function (strategy, selector, elementId
+  shouldCheckStaleness = false,
+  multiple = false,
+} = {}) {
+  const threshold = this.settings.getSettings().imageMatchThreshold;
```
does getSettings return a promise?

mykola-mokhnach replied:
```
const {
  imageMatchThreshold: threshold,
  fixImageTemplateSize
} = this.settings.getSettings();
```

jlipps replied:
> does getSettings return a promise?

no

mykola-mokhnach commented on lib/basedriver/commands/find.js:
```diff
@@ -86,29 +107,49 @@ helpers.findByImage = async function (b64Template, {checkingStaleness = false} =
+    // _other_ kind of error, it means something blew up totally apart from the
+    // implicit wait timeout. We should not mask that error and instead throw
+    // it straightaway
     if (!err.match(/Condition unmet/)) {
```
should me match err.message instead?

jlipps replied:
yes!

mykola-mokhnach commented on lib/basedriver/device-settings.js:
```diff
@@ -3,7 +3,30 @@ import log from './logger';
+
+  // which method to use for tapping by coordinate for image elements. the
+  // options are 'w3c' or 'mjsonwp'
+  imageElementTapStrategy: 'w3c',
```
shouldn't this be detected automatically?

jlipps replied:
how can it be detected automatically? we don't know which functionality the downstream driver supports. i.e., this doesn't (and shouldn't) correspond to the protocol used. it corresponds to the functionality available in the driver.

mykola-mokhnach replied:
we know in which mode the downstream driver is running (w3c/jsonwp), don't we?

jlipps replied:
I'm not sure that we do. And anyway the mode doesn't matter: all we're doing is calling `driver.performActions` or `driver.performTouch`. There's no protocol involved. I'm just giving access to these methods protocol-related names because that's how they're known. But maybe strings like `w3cAction` and `touchAction` would be better.

But your comment makes me realize that this will not be proxied so in the case of WDA, we'll have to ensure the xcuitest driver has `performActions` on it, and manually proxy the command forward to WDA. That's not ideal.

mykola-mokhnach commented on lib/protocol/protocol.js:
```diff
@@ -439,7 +439,7 @@ function driverShouldDoJwpProxy (driver, req, command) {
 
   // if it looks like we have an image element in the body, never proxy
   const stringBody = JSON.stringify(req.body);
```
I'm still not quite sure about that. What if any other valid request includes accidentally includes '-image' substring in its body?

jlipps replied:
it would have to include `appium-image-element-`. but you are right it is a risk. for a similar reason the webdriver spec does the whole `element-6066-11e4-a52e-4f735466cecf` thing.

one alternative is actually walking through the body recursively and seeing if we have an object with the exact form of an element. that seemed like it might be overkill. @imurchie what do you think?

mykola-mokhnach replied:
having such suffix might be useful IMHO

mykola-mokhnach commented on lib/basedriver/image-element.js:
```diff
@@ -6,76 +6,174 @@ import UUID from 'uuid-js';
+        actions: [
+          {type: 'pointerMove', x, y},
+          {type: 'pointerDown'},
+          {type: 'pause', duration: 500},
```
500 is the duration of a long tap. single tap duration is about 100-125 ms

mykola-mokhnach commented on lib/protocol/protocol.js:
```diff
@@ -418,22 +419,14 @@ function driverShouldDoJwpProxy (driver, req, command) {
+
+  // if it looks like we have an image element in the body, never proxy
+  const stringBody = JSON.stringify(req.body);
+  if (stringBody && stringBody.includes(IMAGE_ELEMENT_PREFIX)) {
```
wouldn't it be safer to change includes to `/`\b${IMAGE_ELEMENT_PREFIX}`/.test(stringBody)` ?

jlipps commented:
ok i've fixed all the issues reported so far and written all the unit tests that need writing. i will move on to writing functional tests in the xcui and uiauto2 drivers, so let's wait to merge this until i've at least run them locally. meanwhile i added a few significant commits since the last review so might want to have another look at it.

mykola-mokhnach commented on lib/protocol/protocol.js:
```diff
@@ -17,6 +17,16 @@ const LOG_OBJ_LENGTH = 1024; // MAX LENGTH Logged to file / console
+const IMAGE_ELEMENT_PREFIX = 'appium-image-element-';
+
+const IMG_EL_BODY_RE = new RegExp(
+  `"(${W3C_ELEMENT_KEY}|${MJSONWP_ELEMENT_KEY})": ?` +
```
perhaps ` ?` -> `\s*`

jlipps replied:
good idea, will update

mykola-mokhnach commented on lib/basedriver/driver.js:
```diff
@@ -408,6 +423,31 @@ class BaseDriver extends Protocol {
 
+  proxyRouteIsAvoided (sessionId, method, url) {
+    for (let avoidSchema of this.getProxyAvoidList(sessionId)) {
+      if (!_.isArray(avoidSchema) || avoidSchema.length !== 2) {
```
this will throw inside avoidSchema.length if `avoidSchema` is null

jlipps replied:
existing code (moved from protocol) but good point, i'll fix

jlipps replied:
actually, this is OK. if avoidSchema is not an array, the condition will short-circuit since the LHS is true, and `avoidSchema.length` will never be evaluated

mykola-mokhnach commented on lib/basedriver/driver.js:
```diff
@@ -408,6 +423,31 @@ class BaseDriver extends Protocol {
+      if (!_.includes(['GET', 'POST', 'DELETE'], avoidMethod)) {
+        throw new Error(`Unrecognized proxy avoidance method '${avoidMethod}'`);
+      }
+      if (!(avoidPathRegex instanceof RegExp)) {
```
perhaps `_.isRegExp` ?

jlipps replied:
yep, existing code but will fix

mykola-mokhnach commented on lib/basedriver/driver.js:
```diff
@@ -408,6 +423,31 @@ class BaseDriver extends Protocol {
+    return false;
+  }
+
+  async proxyCommand (/*url, method, body*/) {
```
why do we need to implement it in each driver?

jlipps replied:
each driver does proxying differently (uiauto2 calls `this.uiautomator2.jwproxy` for example, which isn't the case with xcui)

mykola-mokhnach replied:
cannot we just call `click` or `performActions`? Why the base driver should care about proxying?

jlipps replied:
what if a driver doesn't implement `performActions` itself, but merely proxies requests for it to a downstream driver? for example, if uiauto2 server implemented the /actions endpoint, but uiauto2 driver did not. this is a totally valid case. we have many endpoints which are not specified explicitly in a driver and instead we rely on proxying to access the endpoint on a downstream driver.

does that make sense? maybe i'm confused but i think it's a real issue (at least in principle)

mykola-mokhnach replied:
I wouldn't try to outsmart drivers. If some endpoint is not implemented there and it exists in the downstream driver then it would be better and more practical to implement the missing stuff directly there. Adding new requirements to base driver creates unnecessary complexity for the whole infrastructure.

jlipps replied:
ok. i think you have a good point. the way i am doing it now we will also require drivers to implement the `proxyCommand` method. so it's not less onerous than requiring them to simply implement `performActions` and `performTouch`. i'll do the same thing as i have done elsewhere and simply throw a meaningful error if these methods do not exist on the driver.

if the driver typically proxies the action/touch commands, then it will need to basically implement a one-line function to support image elements, like:

```js
commands.performActions = function (actions) {
  this.proxyCommand('POST', '/actions', actions);
};
```

In most cases this will be bypassed, but in the case of ImageElement actions, it will be called.

jlipps commented:
ok, tests have been written for xcuitest driver and uiauto2 driver (those PRs are up but won't pass until this is merged). from my perspective this is solid (at least for a beta, presumably we'll find more bugs or important behavior suggestions when it's in the wild).

jlipps commented:
will also work on documentation, btw.

jlipps commented:
the one build failure is a travis timeout, so I think this is good to go. Any other thoughts @imurchie @mykola-mokhnach? Otherwise I'm ready to merge.

mykola-mokhnach commented on lib/protocol/protocol.js:
```diff
@@ -418,22 +428,26 @@ function driverShouldDoJwpProxy (driver, req, command) {
-    if (avoidMethod === req.method && avoidPathRegex.test(normalizedUrl)) {
-      return false;
-    }
+  if (driver.proxyRouteIsAvoided(req.params.sessionId, req.method, req.originalUrl)) {
```
it might be useful to add a docstring to this new method

jlipps commented:
@mykola-mokhnach ok docstring added. let me know if you're happy with things as they stand now.

jlipps commented:
@imurchie how about you?

jlipps commented:
published in `appium-base-driver@3.4.0`
